File: retired/ts-backend/central-api/src/modules/identity/jwt/jwt-verifier.ts
import type { JwtHeader, TokenSignatureVerifier } from './token-signature-verifier';
import { JwtVerificationError } from './jwt-errors';

export interface VerifiedJwtClaims {
  sub: string;
  iss?: string;
  aud?: string | string[];
  exp?: number;
  nbf?: number;
  iat?: number;
  email?: string;
  name?: string;
  [claim: string]: unknown;
}

export interface VerifiedJwt {
  token: string;
  header: JwtHeader;
  claims: VerifiedJwtClaims;
}

export interface JwtVerifierOptions {
  issuer?: string;
  audience?: string;
  clockSkewSeconds?: number;
  signatureVerifier: TokenSignatureVerifier;
}

interface ParsedJwt {
  header: JwtHeader;
  payload: unknown;
  signingInput: string;
  signatureB64: string;
}

export class JwtVerifier {
  constructor(private readonly options: JwtVerifierOptions) {}

  get algorithm(): 'HS256' | 'RS256' {
    return this.options.signatureVerifier.algorithm;
  }

  get kind(): 'hmac' | 'jwks' | 'unconfigured' {
    return this.options.signatureVerifier.kind;
  }

  async verify(token: string): Promise<VerifiedJwt> {
    if (typeof token !== 'string' || token.length === 0) {
      throw new JwtVerificationError('MALFORMED_TOKEN');
    }

    const parts = this.parse(token);
    if (parts.header.alg !== this.options.signatureVerifier.algorithm) {
      throw new JwtVerificationError('UNSUPPORTED_ALGORITHM', `Unsupported JWT algorithm: ${parts.header.alg}`);
    }

    const signatureValid = await this.options.signatureVerifier.verify(parts);
    if (!signatureValid) {
      throw new JwtVerificationError('SIGNATURE_INVALID');
    }

    const claims = this.validateClaims(parts.payload);

    return {
      token,
      header: { alg: parts.header.alg, kid: parts.header.kid },
      claims,
    };
  }

  private parse(token: string): ParsedJwt {
    const segments = token.split('.');
    if (segments.length !== 3) {
      throw new JwtVerificationError('MALFORMED_TOKEN');
    }

    const [headerB64, payloadB64, signatureB64] = segments;

    const header = this.decodeJson<JwtHeader>(headerB64, 'MALFORMED_TOKEN');
    const payload = this.decodeJson<unknown>(payloadB64, 'MALFORMED_TOKEN');

    return {
      header,
      payload,
      signingInput: `${headerB64}.${payloadB64}`,
      signatureB64,
    };
  }

  private decodeJson<T>(segment: string, reason: 'MALFORMED_TOKEN' | 'MISSING_SUBJECT' = 'MALFORMED_TOKEN'): T {
    try {
      const decoded = Buffer.from(segment, 'base64url').toString('utf8');
      return JSON.parse(decoded) as T;
    } catch (error) {
      throw new JwtVerificationError(reason, `Unable to decode JWT segment: ${String(error)}`);
    }
  }

  private validateClaims(payload: unknown): VerifiedJwtClaims {
    if (typeof payload !== 'object' || payload === null || Array.isArray(payload)) {
      throw new JwtVerificationError('MISSING_SUBJECT');
    }

    const claims = payload as Record<string, unknown>;

    if (typeof claims.sub !== 'string' || claims.sub.length === 0) {
      throw new JwtVerificationError('MISSING_SUBJECT');
    }

    if (this.options.issuer && claims.iss !== this.options.issuer) {
      throw new JwtVerificationError('INVALID_ISSUER');
    }

    if (this.options.audience) {
      const expected = this.options.audience;
      const audienceMatches =
        typeof claims.aud === 'string'
          ? claims.aud === expected
          : Array.isArray(claims.aud) && claims.aud.includes(expected);
      if (!audienceMatches) {
        throw new JwtVerificationError('INVALID_AUDIENCE');
      }
    }

    const now = Math.floor(Date.now() / 1000) - (this.options.clockSkewSeconds ?? 0);

    if (typeof claims.exp === 'number' && claims.exp <= now) {
      throw new JwtVerificationError('EXPIRED_TOKEN');
    }

    if (typeof claims.nbf === 'number' && claims.nbf > now) {
      throw new JwtVerificationError('TOKEN_NOT_YET_VALID');
    }

    return claims as VerifiedJwtClaims;
  }
}

File: backend/central-api/src/main/java/com/cybelinx/platform/api/security/jwt/JwtVerifier.java
package com.cybelinx.platform.api.security.jwt;

import com.nimbusds.jwt.SignedJWT;
import java.time.Instant;
import java.util.List;
import java.util.Map;

/**
 * Port of {@code JwtVerifier}: strict segment parsing, exact algorithm match, signature
 * verification through the configured {@link SignatureVerifier}, then claim validation
 * (sub/iss/aud/exp/nbf with clock skew).
 */
public final class JwtVerifier {

    private final String issuer;
    private final String audience;
    private final long clockSkewSeconds;
    private final SignatureVerifier signatureVerifier;

    public JwtVerifier(String issuer, String audience, long clockSkewSeconds, SignatureVerifier signatureVerifier) {
        this.issuer = issuer;
        this.audience = audience;
        this.clockSkewSeconds = clockSkewSeconds;
        this.signatureVerifier = signatureVerifier;
    }

    public String algorithm() {
        return signatureVerifier.algorithm();
    }

    public String kind() {
        return signatureVerifier.kind();
    }

    public VerifiedJwt verify(String token) {
        if (token == null || token.isEmpty()) {
            throw new JwtVerificationError(JwtReason.MALFORMED_TOKEN);
        }

        SignedJWT jwt;
        try {
            jwt = SignedJWT.parse(token);
        } catch (java.text.ParseException e) {
            throw new JwtVerificationError(JwtReason.MALFORMED_TOKEN);
        }

        String headerAlg = jwt.getHeader().getAlgorithm() == null ? null : jwt.getHeader().getAlgorithm().getName();
        if (!signatureVerifier.algorithm().equals(headerAlg)) {
            throw new JwtVerificationError(JwtReason.UNSUPPORTED_ALGORITHM, "Unsupported JWT algorithm: " + headerAlg);
        }

        boolean signatureValid = signatureVerifier.verify(jwt);
        if (!signatureValid) {
            throw new JwtVerificationError(JwtReason.SIGNATURE_INVALID);
        }

        Map<String, Object> claims = validateClaims(jwt);
        return new VerifiedJwt(token, headerAlg, jwt.getHeader().getKeyID(), claims);
    }

    private Map<String, Object> validateClaims(SignedJWT jwt) {
        Map<String, Object> claims;
        try {
            claims = jwt.getPayload().toJSONObject();
        } catch (Exception e) {
            claims = null;
        }
        if (claims == null || claims.isEmpty()) {
            throw new JwtVerificationError(JwtReason.MISSING_SUBJECT);
        }

        Object sub = claims.get("sub");
        if (!(sub instanceof String subject) || subject.isEmpty()) {
            throw new JwtVerificationError(JwtReason.MISSING_SUBJECT);
        }

        if (issuer != null && !issuer.isEmpty() && !issuer.equals(claims.get("iss"))) {
            throw new JwtVerificationError(JwtReason.INVALID_ISSUER);
        }

        if (audience != null && !audience.isEmpty()) {
            Object aud = claims.get("aud");
            boolean audienceMatches = aud instanceof String s
                    ? audience.equals(s)
                    : aud instanceof List<?> list && list.contains(audience);
            if (!audienceMatches) {
                throw new JwtVerificationError(JwtReason.INVALID_AUDIENCE);
            }
        }

        long now = Instant.now().getEpochSecond() - clockSkewSeconds;

        Object exp = claims.get("exp");
        if (exp instanceof Number expNumber && expNumber.longValue() <= now) {
            throw new JwtVerificationError(JwtReason.EXPIRED_TOKEN);
        }

        Object nbf = claims.get("nbf");
        if (nbf instanceof Number nbfNumber && nbfNumber.longValue() > now) {
            throw new JwtVerificationError(JwtReason.TOKEN_NOT_YET_VALID);
        }

        return claims;
    }

    /** Port of {@code VerifiedJwt}. */
    public record VerifiedJwt(String token, String alg, String kid, Map<String, Object> claims) {}
}